### ISSUE TITLE
Examples dont work on http://localhost:8080 due to incorrect path to pyscript.js build folder

### ISSUE THREAD

**Issue by sumahadevan** [pyscript/pyscript#688]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a solution or duplication
- [X] I already searched in Google and didn't find any good information or help

### What happened?

Platforms: Both Ubuntu 22.95 LTS and macOS Monterey

What happened: After installing npm package rollup globally, cloning the repository, executing 'make setup', 'make examples' and 'npm run dev', the examples start page is displayed at localhost:8080. However on clicking on any example, the page for the example shows only the Python code but doesn't execute it. For example, the page localhost:8080/bokeh.html displays the following:

`- bokeh - numpy
Bokeh Example
import json import pyodide from js import Bokeh, console, JSON from bokeh.embed import json_item from bokeh.plotting import figure from bokeh.resources import CDN # create a new plot with default tools, using figure p = figure(plot_width=400, plot_height=400) # add a circle renderer with x and y coordinates, size, color, and alpha p.circle([1, 2, 3, 4, 5], [6, 7, 2, 4, 5], size=15, line_color="navy", fill_color="orange", fill_alpha=0.5) p_json = json.dumps(json_item(p, "myplot")) Bokeh.embed.embed_item(JSON.parse(p_json)) 
`

What should happen: It should run the python code for that example.

Looking at the page source shows  that it is looking for the pyscript.js build folder in the wrong place. The relevant line of the page source (taken from view-source:http://localhost:8080/bokeh.html) is:

`
    <script defer src="../build/pyscript.js"></script>
`

Following the link to ../build/pyscript.js gives the following page content:

> 404 Not Found
> /home/mads/MDDev/pyscriptProj/pyscript/pyscriptjs/examples/build/pyscript.js`
> (rollup-plugin-serve)

In other words, it expects the build folder to be under examples, which is easily confirmed - on copying the build folder from ../build to examples/build, ALL the examples run. The reason is also clear - since the ROOT folder being served as localhost:8080 points to 'examples', it is not possible to go up one level with relative path. This issue is not there for the live website as explicit URLs are provided there. One possible solution is to serve the root as being one level higher than examples (with a suitable index.html).  

### What browsers are you seeing the problem on? (if applicable)

Firefox, Chrome

**Comment by sumahadevan** [comment]
Platforms are Ubuntu 22.04 LTS and macOS Monterey, typo above.

**Comment by FabioRosado** [comment]
Can also reproduce this issue in Manjaro 21.3.7.  I will open a PR with a potential solution to fix this for local dev

**Comment by sumahadevan** [comment]
@FabioRosado @marimeireles I spotted a small bug in the fix by Fabio for automatically copying the build folder that resolves this issue - namely, if you run 'make examples' a second time, the new build folder (for the second build) is copied INSIDE the existing build folder (of course it works correctly the first time when the build folder does not exist). The culprit seems to be the format of the copy command used in Fabio's fix:

cp -R ./build/ ./examples/build

The updated fix presumably would be to delete the examples/build folder first, or to use an alternative syntax for the copy command. I haven't logged a new issue as it is related to the current issue (and don't think I have the permissions to re-open it) - so just wanted to bring it to your attention (I spotted it by chance when comparing the dates of files in the main build folder with those in the examples/build folder) after a second build.

**Comment by marimeireles** [comment]
@sumahadevan that's great feedback, thank you!
I think we all missed it, but it makes sense.
Would you be open to write a fix for it? As you said, I believe just running a "make clean" or removing the directory would work.
Cheers and thanks!

**Comment by sumahadevan** [comment]
@marimeireles I'll be happy to take care of the fix.

**Comment by marimeireles** [comment]
Awesome!
Ping me if you get stuck 😄

**Comment by sumahadevan** [comment]
@marimeireles thanks for the offer to help - working alone, I very often struggle a bit initially to get things correctly configured etc. 

Today I have fixed the bug in copying the build folder to correct location under examples, but needed to discuss other related issues before raising a pull request.

1) Firstly, the original error in the copy command does not actually require a clean - I realized that ALMOST exactly the same code is being used a few lines earlier to copy the examples folder itself, and this works correctly multiple times. The corrected version of the "make examples" that also works correctly for the build folder is given below, and I have tested it works correctly on both the first and later builds.

`examples:
	mkdir -p ./examples
	cp -r ../examples/* ./examples
	chmod -R 755 examples
	find ./examples/toga -type f -name '*.html' -exec sed $(SED_I_ARG) s+https://pyscript.net/latest/+../../build/+g {} \;
	find ./examples/webgl -type f -name '*.html' -exec sed $(SED_I_ARG) s+https://pyscript.net/latest/+../../../build/+g {} \;
	find ./examples -type f -name '*.html' -exec sed $(SED_I_ARG) s+https://pyscript.net/latest/+../build/+g {} \;
	npm run build
	mkdir -p ./examples/build
	cp -R ./build/* ./examples/build
	@echo "To serve examples run: $(conda_run) python -m http.server 8080 --directory examples"
`

2) Although I am all set to raise a pull request, there are still some other issues. The above assumes that no files from any previous builds are ever discarded (for whatever reason). So in fact BOTH the copy of the examples folder AND the copy of the build folder require a proper preliminary CLEAN. Wanted to confirm this understanding with you before fixing the issue for BOTH the copy commands (not just the second one).

3) I also wanted to remark on something else which I believe can be easily fixed (of course as a separate PR). Namely, currently one cannot build on Windows because it cannot understand the "make" command. But elsewhere in this repository we are also using npm scripts such as "npm run dev". I looked at the Makefile, and actually it is not a true Makefile in the older C/C++ sense where file dates are carefully compared. So switching it to npm scripts is easy, and the main issue is handling the Unix syntax (for example, the "cp -R"). Again, there are npm packages such as **shx** to help with that.

https://www.npmjs.com/package/shx

I estimate it will take only a few days (at most a week) to get the builds to work on Windows (and I have plenty of free time...). I would also like to note that Anaconda itself installs without any problems on my Windows machines, and here (in make setup) we are simply creating another conda environment. I would like your thoughts on this (and I can see a lot of value in bringing Windows into the picture earlier rather than later).

**Comment by sumahadevan** [comment]
Repasting the code in 1) above as it got reformatted.

examples:
	mkdir -p ./examples
	cp -r ../examples/* ./examples
	chmod -R 755 examples
	find ./examples/toga -type f -name '*.html' -exec sed $(SED_I_ARG) s+https://pyscript.net/latest/+../../build/+g {} \;
	find ./examples/webgl -type f -name '*.html' -exec sed $(SED_I_ARG) s+https://pyscript.net/latest/+../../../build/+g {} \;
	find ./examples -type f -name '*.html' -exec sed $(SED_I_ARG) s+https://pyscript.net/latest/+../build/+g {} \;
	npm run build
	mkdir -p ./examples/build
	cp -R ./build/* ./examples/build
	@echo "To serve examples run: $(conda_run) python -m http.server 8080 --directory examples"

**Comment by sumahadevan** [comment]
I have now also added a line in the Makefile to automatically remove the pyscriptjs/examples/build folder each time.

After some thought I decided not to automatically remove the pyscriptjs/examples folder each time, as this folder is often used for work-in-progress when creating new examples (I myself created several examples to test another issue a couple of days ago and realized they would all have been deleted in this examples folder had been automatically removed - so only the build sub-folder is automatically removed before copying a fresh build).

@marimeireles today morning I saw your new issue #770 regarding how to generate the PR. This was very helpful as I was just getting ready to submit my PR (I am aware of the general steps for forking and creating a PR, but not the specific requirements for this repository). In particular, before generating a pull request, I wanted to clarify:

1) whether I should necessarily be using a new branch in my forked repository (although in this case only a single file has been modified, I understand that this would be good practice in general). Please advise.

2) I will also run the pre-commit tool for linting, and do 'make tests' to make sure that nothing is broken.

**Comment by FabioRosado** [comment]
> I have now also added a line in the Makefile to automatically remove the pyscriptjs/examples/build folder each time.
> 
> After some thought I decided not to automatically remove the pyscriptjs/examples folder each time, as this folder is often used for work-in-progress when creating new examples (I myself created several examples to test another issue a couple of days ago and realized they would all have been deleted in this examples folder had been automatically removed - so only the build sub-folder is automatically removed before copying a fresh build).
> 
> @marimeireles today morning I saw your new issue #770 regarding how to generate the PR. This was very helpful as I was just getting ready to submit my PR (I am aware of the general steps for forking and creating a PR, but not the specific requirements for this repository). In particular, before generating a pull request, I wanted to clarify:
> 
> 1. whether I should necessarily be using a new branch in my forked repository (although in this case only a single file has been modified, I understand that this would be good practice in general). Please advise.
> 2. I will also run the pre-commit tool for linting, and do 'make tests' to make sure that nothing is broken.

Hello @sumahadevan  :wave: usually is best practice to create a specific branch for each change. As a maintainer of other projects, using branches makes it easier to checkout the changes and even push new commits if needed. If you use main when I pull your changes it will override my main so then I need to fix that.

Hope this helps!  Let me know if you need help with pre-commit or anything else 👍

**Comment by sumahadevan** [comment]
@FabioRosado yes, this is vey helpful and I understand how using a new branch allows you to separately review my changes in that branch, and even further correct them if needed. I will create a new branch with my changes before submitting the PR.

Regarding pre-commit, I do need one piece of information. I located the .pre-commit-config.yaml file in the root folder, but when I run

pre-commit --version

it says that pre-commit is not found (I thought it might automatically get installed as part of the project). Do I have to install it manually, or is it that I have not yet run some make step for this? Also, if it is to be manually installed, what is the preferred method for this?

**Comment by FabioRosado** [comment]
> @FabioRosado yes, this is vey helpful and I understand how using a new branch allows you to separately review my changes in that branch, and even further correct them if needed. I will create a new branch with my changes before submitting the PR.
> 
> Regarding pre-commit, I do need one piece of information. I located the .pre-commit-config.yaml file in the root folder, but when I run
> 
> pre-commit --version
> 
> it says that pre-commit is not found (I thought it might automatically get installed as part of the project). Do I have to install it manually, or is it that I have not yet run some make step for this? Also, if it is to be manually installed, what is the preferred method for this?

Have you activated your conda environment? When you run `make setup` we create a conda environment at `env/` you can activate it by running `conda activate env/` or `conda activate pyscript/env/` if you are in the root folder. 

If you haven't yet, can you try activating the env and run `pre-commit` after? If it works with the environment activated, you will need to install it with  `pre-commit install`. Then when you commit a new change to your local branch, it will install a few dependencies and run the hooks. (Apologies if you already know all of this!)

**Comment by sumahadevan** [comment]
Yes, that works beautifully. Activated the conda environment and the `pre-commit` command was available. I then ran 'pre-commit install' to setup the pre-commit hook.

Thanks for giving the details! (I haven't used this tool before and was following the instructions on its web page so your guidance was invaluable.) I will now do the necessary for the new branch for the pull request.

**Comment by FabioRosado** [comment]
That's great news! I'm glad I was able to help you with this 😄

**Comment by marimeireles** [comment]
Oh sweet!
Sorry I missed this (was sick!) and thank you so much for jumping in @FabioRosado 🙏

>I also wanted to remark on something else which I believe can be easily fixed (of course as a separate PR). Namely, currently one cannot build on Windows because it cannot understand the "make" command. But elsewhere in this repository we are also using npm scripts such as "npm run dev". I looked at the Makefile, and actually it is not a true Makefile in the older C/C++ sense where file dates are carefully compared. So switching it to npm scripts is easy, and the main issue is handling the Unix syntax (for example, the "cp -R"). Again, there are npm packages such as shx to help with that.

Thanks for this. Brought it up to the team cause I have no idea what are the best practices here.
I'll prbb open a new issue on it (or feel free to open one too) and then mark you there to continue this conversation.

>I estimate it will take only a few days (at most a week) to get the builds to work on Windows (and I have plenty of free time...). I would also like to note that Anaconda itself installs without any problems on my Windows machines, and here (in make setup) we are simply creating another conda environment. I would like your thoughts on this (and I can see a lot of value in bringing Windows into the picture earlier rather than later).

That's so cool! And it's great because I have no idea how to do Windows things 😅
I'll let you know once I have an answer!
